test/query.spec.js

```diff
@@ -570,6 +578,99 @@ describe('Query', () => {
     })
   })
 
+  it('stop after finding k closest peers', (done) => {
+    // mock this so we can dial non existing peers
+    dht.switch.dial = (peer, callback) => callback()
+
+    // Sort peers by distance from peerInfos[0]
+    kadUtils.convertPeerId(peerInfos[0].id, (err, peerZeroDhtKey) => {
+      if (err) {
+        return done(err)
+      }
+
+      const peerIds = peerInfos.map(pi => pi.id)
+      kadUtils.sortClosestPeers(peerIds, peerZeroDhtKey, (err, sorted) => {
+        if (err) {
+          return done(err)
+        }
+
+        // Local node has nodes 10, 16 and 18 in k-bucket
+        const initial = [sorted[10], sorted[16], sorted[18]]
+
+        // Should zoom in to peers near target, and then zoom out again until it
+        // has successfully queried 20 peers
+        const topology = {
+          // Local node has nodes 10, 16 and 18 in k-bucket
+          10: [12, 20, 22, 24, 26, 28],
+          16: [14, 18, 20, 22, 24, 26],
+          18: [4, 6, 8, 12, 14, 16],
+
+          26: [24, 28, 30, 38],
+          30: [14, 28],
+          38: [2],
+
+          // Should zoom out from this point, until it has 20 peers
+          2: [13],
+          13: [15],
+          15: [17],
+
+          // Right before we get to 20 peers, it finds some new peers that are
+          // closer than some of the ones it has already queried
+          17: [1, 3, 5, 11],
+          1: [7, 9],
+          9: [19],
+
+          // At this point it's visited 20 (actually more than 20 peers), and
+          // there are no closer peers to be found, so it should stop querying.
+          // Because there are 3 paths, each with a worker queue with
+          // concurrency 3, the exact order in which peers are visited is
+          // unpredictable, so we add a long tail and below we test to make
+          // sure that it never reaches the end of the tail.
+          19: [21],
+          21: [23],
+          23: [25],
+          25: [27],
+          27: [29],
+          29: [31]
+        }
+
+        const peerIndex = (peerId) => sorted.findIndex(p => p === peerId)
+        const peerIdToInfo = (peerId) => peerInfos.find(pi => pi.id === peerId)
+
+        const visited = []
+        const query = (peerId, cb) => {
+          visited.push(peerId)
+          const i = peerIndex(peerId)
+          const closerIndexes = topology[i] || []
+          const closerPeers = closerIndexes.map(j => peerIdToInfo(sorted[j]))
+          setTimeout(() => cb(null, { closerPeers }))
+        }
+
+        const q = new Query(dht, peerInfos[0].id.id, () => query)
+        q.run(initial, (err, res) => {
+          expect(err).to.not.exist()
+
+          // Should query 20 peers and then stop
+          expect(visited.length).to.be.gt(20)
```

Comment:
Based on the current Topology this should be at least 25 right? Assuming I counted correctly we should have all unique values from below. If that's correct, I think we should actually validate that visited actually contains all of those values.

```js
          10: [12, 20, 22, 24, 26, 28],
          16: [14, 18, 20, 22, 24, 26],
          18: [4, 6, 8, 12, 14, 16],

          26: [24, 28, 30, 38],
          30: [14, 28],
          38: [2],

          // Should zoom out from this point, until it has 20 peers
          2: [13],
          13: [15],
          15: [17],

          // Right before we get to 20 peers, it finds some new peers that are
          // closer than some of the ones it has already queried
          17: [1, 3, 5, 11],
          1: [7, 9],
          9: [19],
```

Other than that, I think this test is a big improvement 👍

Reply:
Makes sense 👍 
Fixed